src/trio/_path.py

```diff
@@ -314,172 +187,78 @@ async def open(  # type: ignore[misc]  # Any usage matches builtins.open().
+    def open(self, *args: Any, **kwargs: Any) -> AsyncIOWrapper[IO[Any]]:
+        return wrap_file(pathlib.Path(self).open(*args, **kwargs))  # noqa: SIM115
+
+    def __repr__(self) -> str:
+        return f"trio.Path({str(self)!r})"
+
+    stat = _wrap_method(pathlib.Path.stat)
+    chmod = _wrap_method(pathlib.Path.chmod)
+    exists = _wrap_method(pathlib.Path.exists)
+    glob = _wrap_method_path_iterable(pathlib.Path.glob)
+    rglob = _wrap_method_path_iterable(pathlib.Path.rglob)
+    is_dir = _wrap_method(pathlib.Path.is_dir)
+    is_file = _wrap_method(pathlib.Path.is_file)
+    is_symlink = _wrap_method(pathlib.Path.is_symlink)
+    is_socket = _wrap_method(pathlib.Path.is_socket)
+    is_fifo = _wrap_method(pathlib.Path.is_fifo)
+    is_block_device = _wrap_method(pathlib.Path.is_block_device)
+    is_char_device = _wrap_method(pathlib.Path.is_char_device)
+    if sys.version_info >= (3, 12):
+        is_junction = _wrap_method(pathlib.Path.is_junction)
+    iterdir = _wrap_method_path_iterable(pathlib.Path.iterdir)
+    lchmod = _wrap_method(pathlib.Path.lchmod)
+    lstat = _wrap_method(pathlib.Path.lstat)
+    mkdir = _wrap_method(pathlib.Path.mkdir)
+    if sys.platform != "win32":
+        owner = _wrap_method(pathlib.Path.owner)
+        group = _wrap_method(pathlib.Path.group)
+    if sys.platform != "win32" or sys.version_info >= (3, 12):
+        is_mount = _wrap_method(pathlib.Path.is_mount)
+    if sys.version_info >= (3, 9):
+        readlink = _wrap_method_path(pathlib.Path.readlink)
+    rename = _wrap_method_path(pathlib.Path.rename)
+    replace = _wrap_method_path(pathlib.Path.replace)
+    resolve = _wrap_method_path(pathlib.Path.resolve)
+    rmdir = _wrap_method(pathlib.Path.rmdir)
+    symlink_to = _wrap_method(pathlib.Path.symlink_to)
+    if sys.version_info >= (3, 10):
+        hardlink_to = _wrap_method(pathlib.Path.hardlink_to)
+    touch = _wrap_method(pathlib.Path.touch)
+    unlink = _wrap_method(pathlib.Path.unlink)
+    absolute = _wrap_method_path(pathlib.Path.absolute)
+    expanduser = _wrap_method_path(pathlib.Path.expanduser)
+    read_bytes = _wrap_method(pathlib.Path.read_bytes)
+    read_text = _wrap_method(pathlib.Path.read_text)
+    samefile = _wrap_method(pathlib.Path.samefile)
+    write_bytes = _wrap_method(pathlib.Path.write_bytes)
+    write_text = _wrap_method(pathlib.Path.write_text)
+    if sys.version_info < (3, 12):
+        link_to = _wrap_method(pathlib.Path.link_to)
+
+
+@final
+class PosixPath(Path, pathlib.PurePosixPath):
+    """A :class:`pathlib.PosixPath` wrapper that executes blocking methods in
+    :meth:`trio.to_thread.run_sync`.
+
+    """
+
+    __slots__ = ()
+
+
+@final
+class WindowsPath(Path, pathlib.PureWindowsPath):
+    """A :class:`pathlib.WindowsPath` wrapper that executes blocking methods in
+    :meth:`trio.to_thread.run_sync`.
+
+    """
+
+    __slots__ = ()
+
+
+final(Path)
```

Comment:
Putting `final` down here prevents type checkers from seeing that `Path` is final. Maybe make an intermediate class? I'm not sure if `Path = final(BasePath)` works but something should work.

Reply:
Oh I was thinking something like:

```py
class PathBase:  # not exported
  # current `Path`

@final
class Path(PathBase):
  # just forward everything to `PathBase`.
  __slots__ = ()

@final
class WindowsPath(PathBase):
  __slots__ = ()
  
  ...
  
```

It's a bit convoluted and just making sure people get earlier warnings about subclassing `trio.Path` might not be worth this.

Reply:
I think I've turned this into a feature now (see latest commit and comment).